nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -842,25 +890,77 @@ ngx_http_modsecurity_detection_task_offload(ngx_http_request_t *r)
 }
 
 static void
-ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_body_handler(ngx_http_request_t* r)
 {
     r->main->count--;
     ngx_http_core_run_phases(r);
 }
 
+static ngx_int_t
+ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_time_t *start_time,
+                                  ngx_http_modsecurity_loc_conf_t * cf)
+{
+    ngx_time_t *end_time;
+    ngx_msec_int_t   ms;
+    ngx_int_t ret;
+    ngx_time_update();
+    end_time = ngx_timeofday();
+
+    ms = (ngx_msec_int_t)
+        ((end_time->sec - start_time->sec) * 1000 + (end_time->msec - start_time->msec));
+    ms = ngx_max(ms, 0);
+
+    ngx_http_variable_value_t* modsec_latency_detect_var = ngx_http_get_indexed_variable(r, modsec_latency_detect_index);
+    ngx_http_variable_value_t* modsec_latency_prevent_var = ngx_http_get_indexed_variable(r, modsec_latency_prevent_index);
+    if (cf->config->is_enabled == MODSEC_DETECTION_ONLY) {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, 0);
+    }
```

Comment:
Ah.. we are setting the other one to 0. Makes sense. Could you please confirm what happens when an unset variable is accessed in the nginx configuration? 
However, what is the reason for having two variables? I was expecting that we are creating two variables because they will probably be setup up differently and thus the need for differentiation. In detection mode how is the asynchronicity being accounted for in the return of the same function handler?

Reply:
Unset variable can have any garbage value and that's why to be sure we dont have any previous or garbage values, I am initializing the value. Also the set function has the memory allocation for the variable as well.
For detection mode we since it is async call, all that matters to us is how much time we blocked/held the nginx request and that is what we are calculating here.  We return NGX_DECLINED in case of detection mode so that we dont wait for the detection thread to complete.